src/mito2/src/worker/handle_catchup.rs

```diff
@@ -0,0 +1,99 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Handling catchup request.
+
+use std::sync::Arc;
+
+use snafu::ensure;
+use store_api::logstore::LogStore;
+use store_api::region_request::{AffectedRows, RegionCatchupRequest};
+use store_api::storage::RegionId;
+
+use crate::error::{self, Result};
+use crate::region::opener::{replay_memtable, RegionOpener};
+use crate::worker::RegionWorkerLoop;
+
+impl<S: LogStore> RegionWorkerLoop<S> {
+    pub(crate) async fn handle_catchup_request(
+        &mut self,
+        region_id: RegionId,
+        request: RegionCatchupRequest,
+    ) -> Result<AffectedRows> {
+        let Some(region) = self.regions.get_region(region_id) else {
+            return error::RegionNotFoundSnafu { region_id }.fail();
+        };
+
+        if region.is_writable() {
+            return Ok(0);
+        }
+        // Note: Currently, We protect the split brain by ensuring the mutable table is empty.
+        // It's expensive to execute catch-up requests without `set_writable=true` multiple times.
+        let is_mutable_empty = region.version().memtables.mutable.is_empty();
+
+        // Utilizes the short circuit evaluation.
+        let region = if !is_mutable_empty || region.manifest_manager.has_update().await? {
+            let reopened_region = Arc::new(
+                RegionOpener::new(
+                    region_id,
+                    region.region_dir(),
+                    self.memtable_builder.clone(),
+                    self.object_store_manager.clone(),
+                    self.scheduler.clone(),
+                )
+                .cache(Some(self.cache_manager.clone()))
+                .options(region.version().options.clone())
+                .skip_wal_replay(true)
+                .open(&self.config, &self.wal)
+                .await?,
+            );
+            debug_assert!(!reopened_region.is_writable());
+            self.regions.insert_region(reopened_region.clone());
+
+            reopened_region
+        } else {
+            region
+        };
+
+        let flushed_entry_id = region.version_control.current().last_entry_id;
+
+        let last_entry_id = replay_memtable(
+            &self.wal,
+            &region.wal_options,
+            region_id,
+            flushed_entry_id,
+            &region.version_control,
+        )
+        .await?;
+
+        let expected_last_entry_id = request.entry_id;
+
+        if let Some(expected_last_entry_id) = expected_last_entry_id {
+            ensure!(
+                expected_last_entry_id == last_entry_id,
+                error::UnexpectedReplaySnafu {
+                    region_id,
+                    expected_last_entry_id,
+                    replayed_last_entry_id: last_entry_id,
+                }
+            )
+        }
```

Comment:
I prefer not to add a new line after each statement. The following code blocks are relevant.
```suggestion
        let flushed_entry_id = region.version_control.current().last_entry_id;
        let last_entry_id = replay_memtable(
            &self.wal,
            &region.wal_options,
            region_id,
            flushed_entry_id,
            &region.version_control,
        )
        .await?;
        if let Some(expected_last_entry_id) = request.entry_id {
            ensure!(
                expected_last_entry_id == last_entry_id,
                error::UnexpectedReplaySnafu {
                    region_id,
                    expected_last_entry_id,
                    replayed_last_entry_id: last_entry_id,
                }
            )
        }
```